### PR TITLE
Authorship

### PR DESCRIPTION
**Pull request by cphyc**
:warning: I didn't add the compiled files in this PR.

### CODE REVIEW

**Comment by cphyc** [comment]
You can now have a name and change it using an (ugly) input box.
It closes #90.

**Comment by cphyc** [comment]
And you can now have an automatically generated color :)
We should however generate the color on the side of the new user, then add it to a `color` shared object so that everybody has the same color.

Closes #84 .

**Comment by cphyc** [comment]
You can now pick a color and it will be applied to you name immediately. The colors are now a shared object.

File: examples/quilljs/index.js [hunk 1]
```diff
@@ -1,69 +1,153 @@
...
+    quill_cursor = quill.getSelection().start;
+    return cursor === quill_cursor;
 }
 
-quill.on("text-change", function(){
-  window.setTimeout(function(){
-    if(editor != null && editor.getDelta != null){
-      console.log("Quill & y-richtext are equal: "+checkConsistency())
+// quill.on("text-change", function(){
+//     if(editor != null && editor.getDelta != null){
```
Comment (dmonad) [inline examples/quilljs/index.js]: why did you remove window.setTimeout? This is necessary for testing!

File: lib/y-richtext.coffee [hunk 2]
```diff
@@ -2,6 +2,9 @@ misc = (require "./misc.coffee")
 BaseClass = misc.BaseClass
 Locker = misc.Locker
 Editors = (require "./editors.coffee")
+Picker = (require "../../quill/src/lib/picker")
+ColorPicker = (require "../../quill/src/lib/color-picker")
```
Comment (dmonad) [inline lib/y-richtext.coffee]: don't you think it is possible to outsource this? I'd rather have y-richtext only for richtext (with binding to ot-types).

File: lib/y-richtext.coffee [hunk 3]
```diff
@@ -253,12 +310,15 @@ class YRichText extends BaseClass
...
-    thisObj._model.getContent("characters").insertContents position, content_array
+    thisObj._model.getContent("characters").insertContents(
+      position, content_array)
 
   deleteHelper = (thisObj, position, length = 1) ->
     thisObj._model.getContent("characters").delete position, length
 
 if window?
+  window.Picker = Picker
+  window.ColorPicker = ColorPicker
```
Comment (dmonad) [inline lib/y-richtext.coffee]: Please, don't pollute the window object! Every Application could have a "Picker" or a "ColorPicker". Then _they_ run into troubles and have to debug it. If you have to, you could put it on Y.RichText (e.g. Y.RichText.ColorPicker). But I currently see no reason to keep the ColorPicker in this file

File: examples/quilljs/index.js [hunk 4]
```diff
@@ -1,69 +1,153 @@
 if(Quill == null){
-  alert("You must download quilljs! It needs to be in the same directory as y-richtext!");
+    alert("You must download quilljs! It needs to be in the same directory as y-richtext!");
```
Comment (dmonad) [inline examples/quilljs/index.js]: Lets agree on two spaces, ok? We also use two spaces in coffeescript. I recently started to work on the Yjs wiki. There I wrote a short article about [code conventions](https://github.com/y-js/yjs/wiki/Code-Conventions)

Comment (cphyc) [inline examples/quilljs/index.js]: Ok.

**Comment by cphyc** [comment]
I fixed your comments and rebased.

File: examples/quilljs/index.js [hunk 5]
```diff
@@ -1,69 +1,195 @@
...
+    d,
+    delta, name, quill_value, value, n;
+  for (d in deltas) {
+    delta = deltas[d];
+    for (name in delta) {
+      value = delta[name];
+      quill_value = quill_deltas[d][name];
+      if (value.constructor === Object) {
+        for (n in value) {
+          if (value[n] !== quill_value[n]) {
```
Comment (dmonad) [inline examples/quilljs/index.js:59]: It is supposed to be `!=` not `!==`. As I wrote in the description you deleted .. 
I need this to check consistency. Otherwise the test will fail if you insert an image. Since I wrote that so explicitely, why did you remove it?

File: examples/quilljs/index.js [hunk 6]
```diff
@@ -1,69 +1,195 @@
...
+            return false;
           }
         }
-      } else if(value != quill_value){
-        return false
+      } else if (value !== quill_value) {
```
Comment (dmonad) [inline examples/quilljs/index.js:63]: here too

File: third_party/color_picker/color-picker.coffee [hunk 7]
```diff
@@ -0,0 +1,16 @@
+dom    = require('./dom')
```
Comment (dmonad) [inline third_party/color_picker/color-picker.coffee:1]: **You must never include code without changing the copyright notice!!**

If you must, you can create a git submodule, and then import the sources. But remove everything you took from quill from this repository!

File: lib/y-richtext.coffee [hunk 8]
```diff
@@ -119,14 +128,40 @@ class YRichText extends BaseClass
   # @param position [Integer] the position where to insert it
   setCursor : (position) ->
     @selfCursor = @_model.getContent("characters").ref(position)
+
     @_model.getContent("cursors").val(@_model.HB.getUserId(), @selfCursor)
 
+  setAuthor : (option) ->
```
Comment (dmonad) [inline lib/y-richtext.coffee:134]: From the maintainers perspective, I don't think it makes sense to implement colors, and mappings from userid->color&name in this module. y-richtext is about richtext, nothing more. It should be so general, that it only handles the rich text structure, and transformations to ot operations. Everything else should be outsourced. You can leave it in y-richtext for now, but put everything related to colors & mappings in one dedicated module.

File: examples/quilljs/index.js (hunk 5 above)
Comment (cphyc) [inline examples/quilljs/index.js:59]: The thing is that images cannot be represented as `{insert : "1"}`, since it would really insert the character "1" in quill. I thought that the "==" were misleading in this case, since we expect quill to have _exactly_ the same content as yjs.

File: lib/y-richtext.coffee (hunk 8 above)
Comment (cphyc) [inline lib/y-richtext.coffee:134]: Can you write this as an issue?
